File: 김현주/프로그래머스/Lv2/[1차] 뉴스 클러스터링/[1차] 뉴스 클러스터링.js
function solution(str1, str2) {
    let answer = 0;
    
    //전부 대문자로 변경
    str1 = str1.toUpperCase();
    str2 = str2.toUpperCase();
    
    //분해해서 넣기
    //맵 이용해서 한개씩 나올대마다 맵에 개수로 넣기
    const map1 = new Map();
    const map2 = new Map();
    
    let pattern_eng = /[a-zA-Z]/;	// 알파벳 체크
    
    let str1Size = 0;
    //str1 분해 과정
    for(let i=0;i<str1.length-1;i++) {
        let c1 = str1[i];
        let c2 = str1[i+1];
        if(!pattern_eng.test(c1) || !pattern_eng.test(c2)) {
            continue; //만약 둘 중 하나라도 알파벳이 아니면 pass
        }
        let str = c1+c2;
        map1.set(str,(map1.get(str) || 0) +1);
        str1Size++;
    }
    
    let str2Size = 0;
    //str2 분해 과정
    for(let i=0;i<str2.length-1;i++) {
        let c1 = str2[i];
        let c2 = str2[i+1];
        if(!pattern_eng.test(c1) || !pattern_eng.test(c2)) {
            continue; //만약 둘 중 하나라도 알파벳이 아니면 pass
        }
        let str = c1+c2;
        map2.set(str,(map2.get(str) || 0) +1);
        str2Size++;
    }
    
    //이제 문자열 비교!
    let intersection = 0; //교집합 개수
    //교집합 구하기
    for (let key of map1.keys()) {
        if(map2.has(key)) {
            intersection += Math.min(map1.get(key), map2.get(key));
        }
    } 
    
    //합집합 개수 = 두 집합의 크기 - 교집합 크기
    let union = str1Size + str2Size - intersection;
    
    if(union === 0) { //공집합인 경우 0으로 못 나누니 예외처리
        answer = 65536;
    }
    else {
        answer = Math.floor(intersection / union * 65536);
    }
        
    return answer;
}

File: 김현주/프로그래머스/Lv2/[1차] 뉴스 클러스터링/[1차] 뉴스 클러스터링.cpp
#include <string>
#include <vector>
#include <cctype>
#include <algorithm>
#include <cmath>
#include <unordered_map>

using namespace std;

int solution(string str1, string str2) {
    int answer = 0;
    //전부 대문자로 변경해주는 과정
    transform(str1.begin(), str1.end(), str1.begin(), ::toupper);
    transform(str2.begin(), str2.end(), str2.begin(), ::toupper);
    
    unordered_map<string, int> map1;
    unordered_map<string, int> map2;
    

    int str1Size = 0;
    //str1을 분해해서 넣는 과정
    for(int i=0;i<str1.length()-1; i++) {
        char c1 = str1[i];
        char c2 = str1[i+1];

        //공백, 숫자, 특수문자가 들어간 글자쌍은 무시
        if(isalpha(c1) == 0 || isalpha(c2) == 0) { //만약 알파벳이 아니라면
            continue;//패스...
        }
        string str = "";
        str.push_back(c1);
        str.push_back(c2);
        if(map1.count(str) == 0) {
            map1[str] = 1;
        }
        else {
            map1[str] += 1;
        }
        str1Size++;
    }

    int str2Size = 0;
    //str2을 분해해서 넣는 과정
    for(int i=0;i<str2.length()-1; i++) {
        char c1 = str2[i];
        char c2 = str2[i+1];

        //공백, 숫자, 특수문자가 들어간 글자쌍은 무시
        if(isalpha(c1) == 0 || isalpha(c2) == 0) { //만약 알파벳이 아니라면
            continue;//패스...
        }
        string str = "";
        str.push_back(c1);
        str.push_back(c2);
        if(map2.count(str) == 0) {
            map2[str] = 1;
        }
        else {
            map2[str] += 1;
        }
        str2Size++;
    }

    //이제 문자열 비교!     
    int intersection = 0; //교집합 개수

    //교집합 구하기
    for (auto const &pair: map1) {
        string key = pair.first;
        if(map2.count(key) > 0) {
            intersection += min(map1[key], map2[key]);
        }
    }

    //합집합의 개수 = 두 백터의 전체 크기 합 - 교집합의 크기
    int unionCount = str1Size + str2Size - intersection;

    //합집합이 아예 없어서 공집합일 경우 (0으로 나눌 수 없으니 오류 처리)
    if(unionCount == 0)
        answer = 65536;
    else{
        float result = ((float)intersection / (float)unionCount)* 65536;
        answer = (int)result;
    }
    
    return answer;
}